### PR TITLE
feat: Add FDv2 wire format and protocol handler

### PR DESCRIPTION
**Pull request by beekld**
## Summary

- Introduces the `fdv2` module:
  - wire-format types
  - assembled `ChangeSet` / `Selector` types
  - protocol-handler state machine that folds SSE events into change sets
- Object bodies stay as raw `serde_json::Value`; typed deserialization will live in a later translator.
- Matches the C++ SDK on permissive-parsing choices:
  - `Selector` keeps the deprecated `version` field alongside `state`
  - `Goodbye` and `FDv2Error` use optional fields to tolerate permissively-parsed wire data
- Module is annotated `#[allow(dead_code)]` for now: types are covered by unit tests but not yet reachable from the production data path.

---

> [!NOTE]
> **Low Risk**
> New code is test-only and not on the live flag data path; behavior changes only when FDv2 is integrated later.
> 
> **Overview**
> Adds a new **`fdv2`** module (registered in `lib.rs` but **`#[allow(dead_code)]`** until the streaming data path uses it) for LaunchDarkly **FDv2** streaming.
> 
> **Wire layer** (`wire.rs`): deserializes SSE payloads—`server-intent`, `put-object` / `delete-object`, `payload-transferred`, `error`, `goodbye`—with permissive parsing (unknown intent codes → `Unknown`, optional fields on `Goodbye` / `FDv2Error`). Put bodies stay as **`serde_json::Value`** for a later translator.
> 
> **Protocol** (`protocol.rs`): **`FDv2ProtocolHandler`** folds events into **`ChangeSet`** values (`None` / `Full` / `Partial` plus put/delete ops and a selector from `payload-transferred.state`). It tracks full vs partial transfer intent, allows further put/delete cycles without a new `server-intent` after the first payload, clears buffered changes on server `error` while keeping transfer state, and resets on `goodbye`.
> 
> Unit tests cover wire parsing and protocol edge cases (malformed JSON, unknown events, protocol violations).
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 7c71c5551e921087adb63bd9f00cbfb5492ac347. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: launchdarkly-server-sdk/src/fdv2/wire.rs [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+
+#[derive(Debug, Deserialize)]
+pub(super) struct ServerIntent {
+    pub(super) payloads: Vec<ServerIntentPayload>,
+}
+
+#[derive(Debug, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub(super) struct ServerIntentPayload {
+    pub(super) intent_code: IntentCode,
```
Comment (keelerm84) [inline launchdarkly-server-sdk/src/fdv2/wire.rs:23]: I think we are missing "reason" here.

File: launchdarkly-server-sdk/src/fdv2/wire.rs [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+}
+
+#[derive(Debug, Deserialize)]
+pub(super) struct PutObject {
+    pub(super) version: u64,
+    pub(super) kind: String,
+    pub(super) key: String,
+    pub(super) object: serde_json::Value,
```
Comment (keelerm84) [inline launchdarkly-server-sdk/src/fdv2/wire.rs:32]: What do we think about using serde's raw value? Might save some cycles if we find we need to discard part of the stream.

File: launchdarkly-server-sdk/src/fdv2/wire.rs [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+    pub(super) version: u64,
+    pub(super) kind: String,
+    pub(super) key: String,
+}
+
+#[derive(Debug, Deserialize)]
+pub(super) struct PayloadTransferred {
+    pub(super) state: String,
+    #[serde(default)]
+    pub(super) version: u64,
```
Comment (keelerm84) [inline launchdarkly-server-sdk/src/fdv2/wire.rs]: You can remove this. We were never meant to process that version.

https://github.com/launchdarkly/sdk-specs/pull/169

File: launchdarkly-server-sdk/src/fdv2/wire.rs [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+#[derive(Debug, Deserialize)]
+pub(super) struct FDv2Error {
+    pub(super) id: Option<String>,
+    pub(super) reason: String,
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub(super) enum Selector {
+    Empty,
+    Set { state: String, version: u64 },
```
Comment (keelerm84) [inline launchdarkly-server-sdk/src/fdv2/wire.rs]: Means you also won't need this I don't believe.

File: launchdarkly-server-sdk/src/fdv2/wire.rs [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+}
+
+#[derive(Debug)]
+pub(super) enum FDv2Change {
+    Put {
+        kind: String,
+        key: String,
+        version: u64,
+        object: serde_json::Value,
+    },
```
Comment (keelerm84) [inline launchdarkly-server-sdk/src/fdv2/wire.rs]: Can you do `Put(PutObject)` and `Delete(DeleteObject)`?

File: launchdarkly-server-sdk/src/fdv2/wire.rs [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+                "reason": "payload-missing"
+            }]
+        });
+        let intent: ServerIntent = serde_json::from_value(payload).unwrap();
+        assert_eq!(intent.payloads.len(), 1);
+        assert_eq!(intent.payloads[0].intent_code, IntentCode::XferFull);
+    }
+
+    #[test]
+    fn put_object_keeps_object_as_raw_json() {
```
Comment (keelerm84) [inline launchdarkly-server-sdk/src/fdv2/wire.rs]: This isn't ensuring it is still raw bytes, just that it could parse an arbitrary object type. If we go with the raw value, this test will have to change a bit.

File: launchdarkly-server-sdk/src/fdv2/wire.rs [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+        });
+        let put: PutObject = serde_json::from_value(payload).unwrap();
+        assert_eq!(put.version, 42);
+        assert_eq!(put.kind, "flag");
+        assert_eq!(put.key, "my-flag");
+        assert_eq!(put.object["on"], json!(true));
+    }
+
+    #[test]
+    fn payload_transferred_defaults_version_when_absent() {
```
Comment (keelerm84) [inline launchdarkly-server-sdk/src/fdv2/wire.rs]: Shouldn't need this test after we remove the version.

File: launchdarkly-server-sdk/src/fdv2/wire.rs (hunk 1 above)
Comment (beekld) [reply launchdarkly-server-sdk/src/fdv2/wire.rs:23]: Added reason, id, and target.

File: launchdarkly-server-sdk/src/fdv2/wire.rs (hunk 2 above)
Comment (beekld) [reply launchdarkly-server-sdk/src/fdv2/wire.rs:32]: I prefer keeping this as `Value` to be consistent with how it's written in C++ and Java. I'm open to discussing it further.

File: launchdarkly-server-sdk/src/fdv2/wire.rs (hunk 3 above)
Comment (beekld) [reply launchdarkly-server-sdk/src/fdv2/wire.rs]: Done

File: launchdarkly-server-sdk/src/fdv2/wire.rs (hunk 5 above)
Comment (beekld) [reply launchdarkly-server-sdk/src/fdv2/wire.rs]: Done.

File: launchdarkly-server-sdk/src/fdv2/wire.rs (hunk 6 above)
Comment (beekld) [reply launchdarkly-server-sdk/src/fdv2/wire.rs]: I renamed this to be clearer.

File: launchdarkly-server-sdk/src/fdv2/protocol.rs [hunk 8]
```diff
@@ -0,0 +1,355 @@
...
+                self.state = State::Partial;
+                ProtocolResult::None
+            }
+            IntentCode::None => {
+                self.state = State::Partial;
+                ProtocolResult::ChangeSet(ChangeSet {
+                    kind: ChangeSetKind::None,
+                    changes: Vec::new(),
+                    selector: Selector::Empty,
+                })
```
[the remark this answers is not in the document]
Comment (beekld) [reply launchdarkly-server-sdk/src/fdv2/protocol.rs:90]: This is consistent with C++ and Java and the spec, AFAICT.

File: launchdarkly-server-sdk/src/fdv2/protocol.rs [hunk 9]
```diff
@@ -0,0 +1,355 @@
...
+            Err(e) => {
+                self.reset();
+                return ProtocolResult::Error(ProtocolError::JsonParse(format!(
+                    "could not deserialize server-intent: {e}"
+                )));
+            }
+        };
+        let Some(payload) = intent.payloads.into_iter().next() else {
+            self.reset();
+            return ProtocolResult::None;
```
[the remark this answers is not in the document]
Comment (beekld) [reply launchdarkly-server-sdk/src/fdv2/protocol.rs:72]: The spec is unclear on what to do in this edge error case. The rust behavior is the same as what we went with in C++, although Java produces an error. I think it's fine how it is.

**Comment by beekld** [comment]
I decided to move some of the types to a separate module, to better reflect how the other SDKs are structured.

**Review by keelerm84 — APPROVED** [review]
